File: hybrid_attendance/android/src/main/kotlin/com/hybridattendance/hybrid_attendance/LocationChecker.kt
package com.hybridattendance.hybrid_attendance

import android.content.Context
import android.location.Location
import android.location.LocationManager
import android.os.Looper
import android.util.Log
import com.google.android.gms.location.*
import com.google.android.gms.tasks.Task
import kotlin.math.*

/**
 * Handles location-based attendance verification.
 */
class LocationChecker(private val context: Context) {
    
    companion object {
        private const val TAG = "LocationChecker"
        private const val LOCATION_TIMEOUT_MS = 10000L // 10 seconds
    }
    
    private val fusedLocationClient: FusedLocationProviderClient by lazy {
        LocationServices.getFusedLocationProviderClient(context)
    }
    
    /**
     * Data class representing a location point.
     */
    data class LocationPoint(val latitude: Double, val longitude: Double)
    
    /**
     * Checks if the current location is within the specified radius of any target locations.
     * 
     * @param targetLocations List of target locations to check against
     * @param radiusMeters Radius in meters for location matching
     * @param enableLogging Whether to enable detailed logging
     * @param onResult Callback with result (isWithinRadius, closestDistance, currentLocation)
     */
    fun checkLocationProximity(
        targetLocations: List<LocationPoint>,
        radiusMeters: Int,
        enableLogging: Boolean,
        onResult: (Boolean, Double?, LocationPoint?) -> Unit
    ) {
        if (enableLogging) {
            Log.d(TAG, "Starting location check for ${targetLocations.size} target locations with radius ${radiusMeters}m")
        }
        
        // Check if location services are available
        if (!isLocationEnabled()) {
            if (enableLogging) {
                Log.e(TAG, "Location services are disabled")
            }
            onResult(false, null, null)
            return
        }
        
        getCurrentLocation(enableLogging) { currentLocation ->
            if (currentLocation == null) {
                if (enableLogging) {
                    Log.e(TAG, "Failed to get current location")
                }
                onResult(false, null, null)
                return@getCurrentLocation
            }
            
            if (enableLogging) {
                Log.d(TAG, "Current location: ${currentLocation.latitude}, ${currentLocation.longitude}")
            }
            
            var closestDistance = Double.MAX_VALUE
            var isWithinRadius = false
            
            for (targetLocation in targetLocations) {
                val distance = calculateDistance(
                    currentLocation.latitude, currentLocation.longitude,
                    targetLocation.latitude, targetLocation.longitude
                )
                
                if (enableLogging) {
                    Log.d(TAG, "Distance to target (${targetLocation.latitude}, ${targetLocation.longitude}): ${distance.toInt()}m")
                }
                
                if (distance < closestDistance) {
                    closestDistance = distance
                }
                
                if (distance <= radiusMeters) {
                    isWithinRadius = true
                    if (enableLogging) {
                        Log.d(TAG, "Location match found within ${distance.toInt()}m")
                    }
                    break
                }
            }
            
            onResult(isWithinRadius, closestDistance, currentLocation)
        }
    }
    
    /**
     * Gets the current location using FusedLocationProviderClient.
     */
    private fun getCurrentLocation(
        enableLogging: Boolean,
        onResult: (LocationPoint?) -> Unit
    ) {
        try {
            val locationRequest = LocationRequest.Builder(Priority.PRIORITY_HIGH_ACCURACY, 1000)
                .setWaitForAccurateLocation(false)
                .setMinUpdateIntervalMillis(500)
                .setMaxUpdateDelayMillis(LOCATION_TIMEOUT_MS)
                .build()
            
            val locationCallback = object : LocationCallback() {
                override fun onLocationResult(locationResult: LocationResult) {
                    super.onLocationResult(locationResult)
                    val location = locationResult.lastLocation
                    if (location != null) {
                        if (enableLogging) {
                            Log.d(TAG, "Received location update: ${location.latitude}, ${location.longitude}")
                        }
                        fusedLocationClient.removeLocationUpdates(this)
                        onResult(LocationPoint(location.latitude, location.longitude))
                    }
                }
                
                override fun onLocationAvailability(locationAvailability: LocationAvailability) {
                    super.onLocationAvailability(locationAvailability)
                    if (!locationAvailability.isLocationAvailable) {
                        if (enableLogging) {
                            Log.w(TAG, "Location not available")
                        }
                    }
                }
            }
            
            // Try to get last known location first
            fusedLocationClient.lastLocation.addOnSuccessListener { location ->
                if (location != null) {
                    val age = System.currentTimeMillis() - location.time
                    if (age < 60000) { // Use if less than 1 minute old
                        if (enableLogging) {
                            Log.d(TAG, "Using cached location (${age}ms old)")
                        }
                        onResult(LocationPoint(location.latitude, location.longitude))
                        return@addOnSuccessListener
                    }
                }
                
                // Request fresh location
                if (enableLogging) {
                    Log.d(TAG, "Requesting fresh location")
                }
                fusedLocationClient.requestLocationUpdates(
                    locationRequest,
                    locationCallback,
                    Looper.getMainLooper()
                )
                
                // Set timeout for location request
                android.os.Handler(Looper.getMainLooper()).postDelayed({
                    fusedLocationClient.removeLocationUpdates(locationCallback)
                    if (enableLogging) {
                        Log.w(TAG, "Location request timeout")
                    }
                    onResult(null)
                }, LOCATION_TIMEOUT_MS)
            }.addOnFailureListener { exception ->
                if (enableLogging) {
                    Log.e(TAG, "Failed to get location: ${exception.message}")
                }
                onResult(null)
            }
            
        } catch (e: SecurityException) {
            if (enableLogging) {
                Log.e(TAG, "Security exception getting location: ${e.message}")
            }
            onResult(null)
        } catch (e: Exception) {
            if (enableLogging) {
                Log.e(TAG, "Exception getting location: ${e.message}")
            }
            onResult(null)
        }
    }
    
    /**
     * Calculates the distance between two geographical points using the Haversine formula.
     * 
     * @return Distance in meters
     */
    private fun calculateDistance(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val earthRadius = 6371000.0 // Earth's radius in meters
        
        val dLat = Math.toRadians(lat2 - lat1)
        val dLon = Math.toRadians(lon2 - lon1)
        
        val a = sin(dLat / 2).pow(2) + cos(Math.toRadians(lat1)) * cos(Math.toRadians(lat2)) * sin(dLon / 2).pow(2)
        val c = 2 * atan2(sqrt(a), sqrt(1 - a))
        
        return earthRadius * c
    }
    
    /**
     * Checks if location services are enabled.
     */
    private fun isLocationEnabled(): Boolean {
        val locationManager = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
        return locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER) ||
               locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER)
    }
}

File: hybrid_attendance/ios/Classes/LocationChecker.swift
import Foundation
import CoreLocation

/// Represents a geographical location point.
struct LocationPoint {
    let latitude: Double
    let longitude: Double
}

/// Handles location-based attendance verification on iOS.
class LocationChecker: NSObject {
    
    private var locationManager: CLLocationManager?
    private var enableLogging = false
    
    // Location request configuration
    private var targetLocations: [LocationPoint] = []
    private var radiusMeters: Int = 100
    
    // Callbacks
    private var onResult: ((Bool, Double?, LocationPoint?) -> Void)?
    
    override init() {
        super.init()
        locationManager = CLLocationManager()
        locationManager?.delegate = self
        locationManager?.desiredAccuracy = kCLLocationAccuracyBest
        locationManager?.distanceFilter = 10 // Update every 10 meters
    }
    
    /// Checks if the current location is within the specified radius of any target locations.
    ///
    /// - Parameters:
    ///   - targetLocations: List of target locations to check against
    ///   - radiusMeters: Radius in meters for location matching
    ///   - enableLogging: Whether to enable detailed logging
    ///   - onResult: Callback with result (isWithinRadius, closestDistance, currentLocation)
    func checkLocationProximity(
        targetLocations: [LocationPoint],
        radiusMeters: Int,
        enableLogging: Bool,
        onResult: @escaping (Bool, Double?, LocationPoint?) -> Void
    ) {
        self.targetLocations = targetLocations
        self.radiusMeters = radiusMeters
        self.enableLogging = enableLogging
        self.onResult = onResult
        
        if enableLogging {
            print("LocationChecker: Starting location check for \(targetLocations.count) target locations with radius \(radiusMeters)m")
        }
        
        guard let locationManager = locationManager else {
            if enableLogging {
                print("LocationChecker: Location manager not available")
            }
            onResult(false, nil, nil)
            return
        }
        
        // Check if location services are enabled
        if !CLLocationManager.locationServicesEnabled() {
            if enableLogging {
                print("LocationChecker: Location services are disabled")
            }
            onResult(false, nil, nil)
            return
        }
        
        // Check authorization status
        let authorizationStatus: CLAuthorizationStatus
        if #available(iOS 14.0, *) {
            authorizationStatus = locationManager.authorizationStatus
        } else {
            authorizationStatus = CLLocationManager.authorizationStatus()
        }
        
        if authorizationStatus != .authorizedWhenInUse && authorizationStatus != .authorizedAlways {
            if enableLogging {
                print("LocationChecker: Location permission not granted, status: \(authorizationStatus.rawValue)")
            }
            onResult(false, nil, nil)
            return
        }
        
        // Try to get cached location first
        if let cachedLocation = locationManager.location {
            let age = Date().timeIntervalSince(cachedLocation.timestamp)
            if age < 60 { // Use if less than 1 minute old
                if enableLogging {
                    print("LocationChecker: Using cached location (\(Int(age))s old)")
                }
                processLocation(cachedLocation)
                return
            }
        }
        
        // Request fresh location
        if enableLogging {
            print("LocationChecker: Requesting fresh location")
        }
        
        locationManager.requestLocation()
        
        // Set timeout for location request
        DispatchQueue.main.asyncAfter(deadline: .now() + 10.0) { [weak self] in
            guard let self = self else { return }
            if self.onResult != nil {
                if self.enableLogging {
                    print("LocationChecker: Location request timeout")
                }
                self.onResult?(false, nil, nil)
                self.onResult = nil
            }
        }
    }
    
    /// Processes the received location and checks proximity to target locations.
    private func processLocation(_ location: CLLocation) {
        let currentLocation = LocationPoint(latitude: location.coordinate.latitude, longitude: location.coordinate.longitude)
        
        if enableLogging {
            print("LocationChecker: Current location: \(currentLocation.latitude), \(currentLocation.longitude)")
        }
        
        var closestDistance = Double.greatestFiniteMagnitude
        var isWithinRadius = false
        
        for targetLocation in targetLocations {
            let targetCLLocation = CLLocation(latitude: targetLocation.latitude, longitude: targetLocation.longitude)
            let distance = location.distance(from: targetCLLocation)
            
            if enableLogging {
                print("LocationChecker: Distance to target (\(targetLocation.latitude), \(targetLocation.longitude)): \(Int(distance))m")
            }
            
            if distance < closestDistance {
                closestDistance = distance
            }
            
            if distance <= Double(radiusMeters) {
                isWithinRadius = true
                if enableLogging {
                    print("LocationChecker: Location match found within \(Int(distance))m")
                }
                break
            }
        }
        
        onResult?(isWithinRadius, closestDistance, currentLocation)
        onResult = nil
    }
}

// MARK: - CLLocationManagerDelegate
extension LocationChecker: CLLocationManagerDelegate {
    
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let location = locations.last, onResult != nil else { return }
        
        if enableLogging {
            print("LocationChecker: Received location update: \(location.coordinate.latitude), \(location.coordinate.longitude)")
        }
        
        processLocation(location)
    }
    
    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        if enableLogging {
            print("LocationChecker: Failed to get location: \(error.localizedDescription)")
        }
        
        if onResult != nil {
            onResult?(false, nil, nil)
            onResult = nil
        }
    }
    
    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        if enableLogging {
            print("LocationChecker: Location authorization changed: \(status.rawValue)")
        }
        
        // If permission is denied during location request, fail immediately
        if onResult != nil && status != .authorizedWhenInUse && status != .authorizedAlways {
            onResult?(false, nil, nil)
            onResult = nil
        }
    }
    
    @available(iOS 14.0, *)
    func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        locationManager(manager, didChangeAuthorization: manager.authorizationStatus)
    }
}
